### ISSUE TITLE
A subunit diagonal carry is not mathematically necessary for convergence

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#20]
## Finding

At commit `af3f866dea23e38c89e88f485bce99ff546d93a9`, the Buehlmaier annotation says:

> “Pre-norm LayerNorm makes the recurrence Jacobian asymptotically contractive at fixed points even where its operator norm exceeds one, **so a diagonal-carry constraint is necessary but not sufficient for convergence**.”

Source: [`bibliography.html` lines 443–448](https://github.com/microprediction/latentreasoning/blob/af3f866dea23e38c89e88f485bce99ff546d93a9/bibliography.html#L443-L448).

The paper provides evidence for **not sufficient**, but neither its analysis nor basic dynamical-systems algebra makes the carry constraint necessary for convergence of the full recurrence.

## Why necessity does not follow

The recurrence is `G(h) = D h + f(h,e)`, so the local Jacobian at a fixed point is `J_G(h*) = D + J_f(h*,e)`.

Local asymptotic stability is governed by the eigenvalues of the **sum**, not by the diagonal term in isolation. A condition such as `spectral_radius(D) < 1` can be a useful design constraint, but it is not mathematically necessary for `spectral_radius(J_G) < 1`, because `J_f` can counteract `D`.

A one-dimensional counterexample is immediate:

```
D = 2
f(h) = -1.5 h
G(h) = 0.5 h
```

Here the carry violates `spectral_radius(D) < 1`, yet every trajectory converges geometrically to the fixed point. More generally, local cancellation or rotation in `J_f` can stabilize a full Jacobian even when a component has spectral radius above one. Conversely, `spectral_radius(D) < 1` does not ensure convergence because the nonlinear term may destabilize the sum.

The paper directly demonstrates the latter direction: 1–2 of 10 small random initializations fail to settle even though the carry is constrained below one. See [§5.3, Table 4](https://arxiv.org/html/2607.10681#S5.SS3). That establishes non-sufficiency.

For necessity, the paper reports a much narrower training ablation: removing the constraint allows gradient descent to push the carry above one at higher learning rates in toy-scale experiments. It does not prove that every convergent pre-LN loop must have `spectral_radius(D) < 1`. The authors themselves scope the empirical work to dimensions 8–32, 2–4 heads, sequence lengths 1–32, and state that larger-scale/richer-task verification is needed; see the [introduction and scope](https://arxiv.org/html/2607.10681#S1).

## Related precision

The paper's full-Jacobian spectral-radius measurements certify **local asymptotic stability at a verified fixed point**, not global convergence from arbitrary initialization. Indeed, some runs never reach the fixed point at which the Jacobian is evaluated. “Asymptotically contractive at every verified fixed point” is therefore an empirical conditional over converged small-scale runs, not a general LayerNorm guarantee.

## Consequence

The current annotation turns an empirically useful safeguard into a necessary condition on the architecture. That can mislead readers into treating a subunit diagonal carry as a theorem-level requirement and ignoring the actual stability object, `D + J_f`, its basin of attraction, and transient non-normal dynamics.

## Concrete correction

A source-faithful replacement would be:

> “In small pre-LN looped transformers, every run that reached a verified fixed point had full-Jacobian spectral radius below one even when its operator norm exceeded one; 1–2 of 10 seeds did not converge despite the diagonal carry being constrained below one. Thus the carry constraint was not sufficient, while ablations suggest it is a useful training safeguard—not a proven necessary condition for convergence. Local stability is governed by the spectral radius of the full Jacobian.”

Primary source: Buehlmaier, [*LayerNorm as Implicit Gain Control in Looped Transformers*](https://arxiv.org/abs/2607.10681).

**Comment by microprediction** [comment]
Correct. Local stability is governed by the spectrum of the full Jacobian, not the carry term alone, so necessity does not follow. The entry now says only that the carry constraint on its own does not guarantee convergence.

Fixed in 0c96d71.

**Comment by microprediction** [comment]
The current fix remains incomplete at commit `ef9bb829b8866cf29902f6093fcbb4b71126d0dc`. [bibliography.html lines 462–466](https://github.com/microprediction/latentreasoning/blob/ef9bb829b8866cf29902f6093fcbb4b71126d0dc/bibliography.html#L462-L466) now says:

> Pre-norm LayerNorm makes the recurrence Jacobian asymptotically contractive at fixed points even where its operator norm exceeds one

That still promotes a conditional small-scale observation into a general causal claim about LayerNorm.

In [Buehlmaier, arXiv:2607.10681v1](https://arxiv.org/html/2607.10681), the subunit spectral radii are empirical: Table 4 reports them only for tiny random models and only among seeds that reached a verified fixed point. The experiments use dimensions 8–32, 2–4 heads, sequence lengths 1–32, and 10 seeds per carry; 1–2 seeds per setting did not settle. The analysis proves bounded normalized increments, fixed-point existence, and inverse-norm scaling of the local increment Jacobian. It does not prove that pre-LayerNorm makes every fixed point attracting; indeed, the paper's operator-norm bound exceeds one except asymptotically.

A direct counterexample shows why the present sentence is false in general. Let (d\ge3), use standard LayerNorm without affine gain/shift (and omit epsilon for clarity), and choose (h_*) with zero mean and unit variance. At (h_*),

[
J_{\mathrm{LN}}(h_*)=P=I-\frac1d\mathbf 1\mathbf 1^\top-\frac1d h_*h_*^\top,
]

the orthogonal projection onto the tangent subspace perpendicular to both (mathbf 1) and (h_*). Define

[
G(h)=\rho h+c\,\mathrm{LN}(h)+b,qquad
0<\rho<1,quad c>1-\rho,quad b=(1-\rho-c)h_*.
]

Then (G(h_*)=h_*), but

[
J_G(h_*)=\rho I+cP.
]

For every nonzero tangent vector (v), (Pv=v), so the corresponding eigenvalue is (
ho+c>1). Thus this pre-LayerNorm recurrence has a locally unstable fixed point despite a subunit diagonal carry. The Jacobian is also symmetric here, so LayerNorm does not inherently make the full Jacobian non-normal. With the usual epsilon, the tangent derivative is merely rescaled by a positive factor near one; choosing (c) larger preserves the counterexample.

Suggested replacement:

> In Buehlmaier's small random pre-LayerNorm models, every run that reached a verified fixed point had full-Jacobian spectral radius below one even when operator norm exceeded one; 1–2 of 10 seeds per setting did not settle. The analysis shows inverse-norm gain control and fixed-point existence, but does not prove that LayerNorm makes arbitrary fixed points attracting. Local stability is governed by the spectrum of the full Jacobian.

This preserves the useful empirical result while separating it from the unsupported general mechanism claim.
